### PR TITLE
ExecutorService instrumentation

### LINKED ISSUE #145: In process context propagation for Executors
To propagate the trace context to different threads, we should by default instrument `ExecutorService`s to propagate the currently active span. The user should not have to manually wrap the `Executor` in order to enable the context propagation for it.

Prior art:
* https://github.com/talsma-ict/context-propagation
* https://github.com/census-instrumentation/opencensus-java/blob/master/contrib/agent/src/main/java/io/opencensus/contrib/agent/instrumentation/ExecutorInstrumentation.java
* https://github.com/census-instrumentation/opencensus-java/blob/master/contrib/agent/src/main/java/io/opencensus/contrib/agent/instrumentation/ThreadInstrumentation.java
* https://github.com/opentracing-contrib/java-concurrent
* https://github.com/openzipkin/brave/blob/3774c2c0681304c4b9e1bf29f701a484b5f65e6f/brave/src/main/java/brave/propagation/CurrentTraceContext.java#L200
* https://github.com/glowroot/glowroot/blob/master/agent/plugins/executor-plugin/src/main/java/org/glowroot/agent/plugin/executor/ExecutorAspect.java

One design decision is whether we want to wrap `Runnable`s  and store the context as an instance variable or if we want to instrument implementations of `Runnable` and add an instance variable via BCI. Wrapping has the disadvantage that we are creating garbage and type casts to a specific sub-type of `Runnable` would fail. The disadvantage of adding the instance variable via BCI is that this approach would not work with runtime attachment of the agent. Yet another option would be to store the trace context in a map which associates Runnables with the corresponding trace context.

Another question is whether we want to generically instrument all runnables or only those submitted to an `Executor`. We could store the trace context when the runnable is created and restore it in the run method. The assumption here is that the `Runnable` is instantiated in the Thread which holds the current trace context. But this assumption is not always true, for example when the `Runnable` is stored in an instance variable.

Relevant interfaces are: `java.util.concurrent.Executor`, `java.util.concurrent.ExecutorService`, `java.util.concurrent.ScheduledExecutorService`, `java.lang.Runnable` and `java.util.concurrent.Callable`

This should fix https://discuss.elastic.co/t/java-apm-agent-sends-no-timeline-spans/141150, where Hystrix is used for fault tolerance of JDBC queries.

### PR DESCRIPTION
**Pull request by felixbarny**
closes elastic/apm-agent-java#145

### CODE REVIEW

File: apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java [hunk 1]
```diff
@@ -80,13 +80,13 @@ private static JsonSchema getSchema(String resource) {
     }
 
     @Override
-    public void report(Transaction transaction) {
+    public synchronized void report(Transaction transaction) {
```
Comment (eyalkoren) [inline apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java]: what's the benefit of synching here? isn't there still a race condition?

Comment (eyalkoren) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java]: Does this change the visibility status? The Transactions/Spans/Errors go into containers, for which the reference doesn't change. I would assume making these collections synchronize be more suitable, but if it works- it works 😄

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContextHolder.java [hunk 2]
```diff
@@ -62,44 +63,30 @@ protected TraceContextHolder(ElasticApmTracer tracer) {
...
-        } catch (Throwable t) {
+        tracer.activate(this);
+        List<ActivationListener> activationListeners = tracer.getActivationListeners();
+        for (int i = 0; i < activationListeners.size(); i++) {
             try {
-                logger.error("Unexpected error while activating context", t);
-            } catch (Throwable ignore) {
+                activationListeners.get(i).onActivate(this);
+            } catch (Error e) {
+                throw e;
```
[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContextHolder.java]: OK, as long as the failure in listeners do not fail the activation I guess it's good.

File: apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java (hunk 1 above)
Comment (felixbarny) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/MockReporter.java]: Right, making the collections synchronized might be better

**Comment by felixbarny** [comment]
I don't think we need to instrument `Future.cancel`. There are three states related to canceling a Future:

* Runnable did not yet run
  This means that the wrapped runnable won't be recycled, but GC will take care of it and nothing bad will happen.
* Runnable is currently running
  The runnable may be interrupted and may throw an exception. In that case, the runnable will still be recycled
* Runnable has already run
   Nothing special to consider here

File: apm-agent-plugins/apm-java-concurrent-plugin/src/main/java/co/elastic/apm/agent/concurrent/ExecutorInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        // which disables context propagation for those
+        // There is a slight risk that retrying causes a side effect but the more likely scenario is that adding the task to the queue
+        // fails and noting has been executed yet.
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Exception.class, repeatOn = Advice.OnNonDefaultValue.class)
+        public static boolean onError(@Advice.This Executor thiz,
+                                      @Nullable @Advice.Thrown Exception exception,
+                                      @Nullable @Advice.Argument(value = 0, readOnly = false) Runnable runnable,
+                                      @Advice.Local("original") @Nullable Runnable original) {
+
+            if (original != null && exception instanceof ClassCastException || exception instanceof IllegalArgumentException) {
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-plugins/apm-java-concurrent-plugin/src/main/java/co/elastic/apm/agent/concurrent/ExecutorInstrumentation.java]: good catch!

File: apm-agent-plugins/apm-java-concurrent-plugin/src/test/java/co/elastic/apm/agent/concurrent/FailingExecutorInstrumentationTest.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Test
+    void testNoInfiniteLoop() {
+        Assertions.assertThatThrownBy(() -> executor.execute(() -> {
+        })).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    // verifies that the Runnables/Callables are executed at most once
+    void failIfCalledTwice() {
+        assertThat(counter.incrementAndGet()).isLessThan(2);
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-plugins/apm-java-concurrent-plugin/src/test/java/co/elastic/apm/agent/concurrent/FailingExecutorInstrumentationTest.java]: JUnit creates a fresh instance of the test class for each test method. I would only have to reset if the counter was static.

[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-plugins/apm-java-concurrent-plugin/src/test/java/co/elastic/apm/agent/concurrent/FailingExecutorInstrumentationTest.java]: wasn't aware of that. I'd still say it makes sense to reset as part of the `@BeforeEach`, like we do with everything else so it is explicit

[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-plugins/apm-java-concurrent-plugin/src/test/java/co/elastic/apm/agent/concurrent/FailingExecutorInstrumentationTest.java]: I'm assigning the AtomicInteger in `@BeforeEach` now :)

**Pull request merged**